File: specops/scripts/utils/io.py
from __future__ import annotations

import json
from pathlib import Path


def resolve_path(input_path: str) -> Path:
    return Path(input_path).expanduser().resolve()


def read_json(file_path: str):
    path = resolve_path(file_path)
    return json.loads(path.read_text(encoding="utf-8"))


def read_text(file_path: str) -> str:
    path = resolve_path(file_path)
    return path.read_text(encoding="utf-8")


def write_text(file_path: str, content: str) -> Path:
    path = resolve_path(file_path)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")
    return path


def apply_template(template: str, replacements: dict[str, str]) -> str:
    rendered = template
    for key, value in replacements.items():
        rendered = rendered.replace(f"{{{{{key}}}}}", str(value))
    return rendered

File: specops/scripts/utils/io.js
const fs = require("fs");
const path = require("path");

function parseArgs(argv) {
  const options = {};
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (!token.startsWith("--")) {
      continue;
    }
    const key = token.slice(2);
    const value = argv[index + 1];
    if (!value || value.startsWith("--")) {
      options[key] = true;
      continue;
    }
    options[key] = value;
    index += 1;
  }
  return options;
}

function resolvePath(inputPath) {
  return path.resolve(process.cwd(), inputPath);
}

function readJson(filePath) {
  const absolutePath = resolvePath(filePath);
  const content = fs.readFileSync(absolutePath, "utf8");
  return JSON.parse(content);
}

function readText(filePath) {
  const absolutePath = resolvePath(filePath);
  return fs.readFileSync(absolutePath, "utf8");
}

function ensureDirectoryFor(filePath) {
  const absolutePath = resolvePath(filePath);
  fs.mkdirSync(path.dirname(absolutePath), { recursive: true });
  return absolutePath;
}

function writeText(filePath, content) {
  const absolutePath = ensureDirectoryFor(filePath);
  fs.writeFileSync(absolutePath, content, "utf8");
  return absolutePath;
}

function applyTemplate(template, replacements) {
  return Object.entries(replacements).reduce(
    (acc, [key, value]) => acc.replace(new RegExp(`{{${key}}}`, "g"), String(value)),
    template,
  );
}

module.exports = {
  applyTemplate,
  parseArgs,
  readJson,
  readText,
  resolvePath,
  writeText,
};